### PR TITLE
Added Qwen2-MoE model, optimizing its performance on Gaudi

### PR DESCRIPTION
**Pull request by gyou2021**
Added Qwen2-MoE model into optimum-habana, optimizing its performance on Gaudi with kv_cache, reuse_cache, flash attention, internal bucket, etc.

Test command:

cd optimum-habana

python -m pytest tests/test_text_generation_example.py -v -s

cd examples/text-generation/

python run_generation.py --model_name_or_path --use_hpu_graphs --use_kv_cache --reuse_cache --trim_logits --attn_softmax_bf16 --use_flash_attention --max_input_tokens 256 --max_new_tokens 256 --batch_size 2 --limit_hpu_graphs --bucket_size 128 --bucket_internal --n_iterations 2 --warmup 2 --bf16

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

### CODE REVIEW

**Comment by gyou2021** [comment]
@ssarkar2 I added Qwen2-MoE test code in test_text_generation_example.py according to your comments in https://github.com/huggingface/optimum-habana/pull/1162. What's more, I created a new branch qwen2-moe following the guidance of https://github.com/huggingface/optimum/blob/main/CONTRIBUTING.md. Thank you.

**Review by imangohari1 — CHANGES_REQUESTED** [review]
Hi @gyou2021 
Thank you for this PR. 
I have gone through this and the the tests seem running fine. 
I made a change in CI tests since `Qwen1.5-7B` was already there, and instead added the Qwen2-7B. 

Please apply the patch with `git am < 0001*` and push. 
We review afterwards. 
Thanks. 
[0001-fea-ci-Added-qwen2-7b-bs-512.patch](https://github.com/user-attachments/files/16970632/0001-fea-ci-Added-qwen2-7b-bs-512.patch)

**Comment by gyou2021** [comment]
Hi @imangohari1 

Thank you for your comments.

I added Qwen2-7B test and removed repeated Qwen1.5-7B test in tests/test_text_generation_example.py according to 0001-fea-ci-Added-qwen2-7b-bs-512.patch. Since "error: tests/test_text_generation_example.py: does not match index" showed when I ran git am < 0001* to apply the patch, I manually modified the code and push it. 

> Hi @gyou2021 Thank you for this PR. I have gone through this and the the tests seem running fine. I made a change in CI tests since `Qwen1.5-7B` was already there, and instead added the Qwen2-7B.
> 
> Please apply the patch with `git am < 0001*` and push. We review afterwards. Thanks. [0001-fea-ci-Added-qwen2-7b-bs-512.patch](https://github.com/user-attachments/files/16970632/0001-fea-ci-Added-qwen2-7b-bs-512.patch)

**Review by imangohari1 — APPROVED** [review]
LGTM! 
I have tested these changes with local cmd and CI. 
they appear functional. 

This PR has also been tested via a Integration branch https://github.com/huggingface/optimum-habana/compare/main...imangohari1:optimum-habana:ig/ci_20240911?expand=1 (CI#255) and they have not introduced a regression. 

@regisss  please take a look when you had a chance.

**Review by regisss — COMMENTED** [review]
There are quite a lot of commented code in the modeling file, can you clean that a bit please?

**Comment by imangohari1** [comment]
@gyou2021 Can you please address Regiss comments? Tnx

File: README.md [hunk 1]
```diff
@@ -211,6 +211,7 @@ The following model architectures, tasks and device distributions have been vali
 | Mixtral |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | Persimmon |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | Qwen2 | <div style="text-align:left"><li>Single card</li></div> | <div style="text-align:left"><li>Single card</li></div> | <li>[language modeling](https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling)</li><li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
+| Qwen2-MoE |   | <div style="text-align:left"><li>Single card</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
```
[the remark this answers is not in the document]
Comment (gyou2021) [reply README.md:214]: > Can you add the same in
> 
> https://github.com/huggingface/optimum-habana/blob/091c8a52655eb35a3feaea93b4b3d98b1f6551fc/docs/source/index.mdx?plain=1#L59
> 
> please?
Added. 
Thank you for your comment.

[the remark this answers is not in the document]
Comment (gyou2021) [reply README.md:214]: > Can you add the same in
> 
> https://github.com/huggingface/optimum-habana/blob/091c8a52655eb35a3feaea93b4b3d98b1f6551fc/docs/source/index.mdx?plain=1#L59
> 
> please?

Added.
Thank you for your comment.

**Comment by gyou2021** [comment]
> There are quite a lot of commented code in the modeling file, can you clean that a bit please?
Thank you for your comment.
The code has been cleaned.

**Comment by gyou2021** [comment]
> @gyou2021 Can you please address Regiss comments? Tnx

Sure. Done.

**Comment by regisss** [comment]
@gyou2021 There are merge conflicts to solve

**Review by regisss — APPROVED** [review]

**Comment by gyou2021** [comment]
> @gyou2021 There are merge conflicts to solve edit: solved

Thank you.
